File: mobile/ios/OTravelz/Features/Offline/WeatherCacheStore.swift
import Foundation

/// Minimal last-known observation cache store for iOS.
/// Formats relative time elapsed without arbitrary hard TTLs.
/// Cached observations are NEVER labeled as live.
public final class WeatherCacheStore: @unchecked Sendable {
    public static let shared = WeatherCacheStore()

    private let defaults: UserDefaults
    private let lock = NSLock()
    private var memoryCache: [String: CachedObservation] = [:]

    public struct CachedObservation: Codable, Sendable {
        public let locationName: String
        public let temperatureC: Double
        public let condition: String
        public let advice: String?
        public let observedAt: Date
    }

    public init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
    }

    private func makeKey(lat: Double, lon: Double) -> String {
        let latRound = String(format: "%.2f", lat)
        let lonRound = String(format: "%.2f", lon)
        return "weather_\(latRound)_\(lonRound)"
    }

    public func saveObservation(
        lat: Double,
        lon: Double,
        locationName: String,
        temperatureC: Double,
        condition: String,
        advice: String?,
        observedAt: Date = Date()
    ) {
        let key = makeKey(lat: lat, lon: lon)
        let observation = CachedObservation(
            locationName: locationName,
            temperatureC: temperatureC,
            condition: condition,
            advice: advice,
            observedAt: observedAt
        )

        lock.lock()
        memoryCache[key] = observation
        lock.unlock()

        if let data = try? JSONEncoder().encode(observation) {
            defaults.set(data, forKey: key)
        }
    }

    public func getCachedObservation(
        lat: Double,
        lon: Double,
        now: Date = Date()
    ) -> WeatherState? {
        let key = makeKey(lat: lat, lon: lon)

        lock.lock()
        let mem = memoryCache[key]
        lock.unlock()

        if let mem = mem {
            let relative = formatRelativeTimeAgo(observedAt: mem.observedAt, now: now)
            return .cached(
                locationName: mem.locationName,
                temperatureC: mem.temperatureC,
                condition: mem.condition,
                advice: mem.advice,
                relativeTimeAgo: relative
            )
        }

        guard let data = defaults.data(forKey: key),
              let obs = try? JSONDecoder().decode(CachedObservation.self, from: data) else {
            return nil
        }

        lock.lock()
        memoryCache[key] = obs
        lock.unlock()

        let relative = formatRelativeTimeAgo(observedAt: obs.observedAt, now: now)
        return .cached(
            locationName: obs.locationName,
            temperatureC: obs.temperatureC,
            condition: obs.condition,
            advice: obs.advice,
            relativeTimeAgo: relative
        )
    }

    public func formatRelativeTimeAgo(observedAt: Date, now: Date) -> String {
        let diffSeconds = max(0, now.timeIntervalSince(observedAt))
        let diffMinutes = Int(diffSeconds / 60)
        let diffHours = diffMinutes / 60
        let diffDays = diffHours / 24

        if diffMinutes < 1 {
            return "just now"
        } else if diffMinutes < 60 {
            return "\(diffMinutes)m"
        } else if diffHours < 24 {
            return "\(diffHours)h"
        } else {
            return "\(diffDays)d"
        }
    }
}

File: mobile/android/src/main/kotlin/com/otravelz/android/offline/WeatherCacheStore.kt
package com.otravelz.android.offline

import android.content.Context
import android.content.SharedPreferences
import com.otravelz.android.data.network.adapter.WeatherState
import java.util.concurrent.ConcurrentHashMap

/**
 * Minimal last-known weather observation cache store.
 * - Stores temperature, condition, advice, and observation timestamp.
 * - Formats relative elapsed time without arbitrary hard TTLs.
 * - Cached observations are NEVER labeled as live.
 */
class WeatherCacheStore private constructor(context: Context?) {

    private val prefs: SharedPreferences? = try {
        context?.getSharedPreferences("otravelz_weather_cache", Context.MODE_PRIVATE)
    } catch (t: Throwable) {
        null
    }

    // In-memory fallback for headless unit tests or fast access
    private val memoryStore = ConcurrentHashMap<String, CachedObservation>()

    data class CachedObservation(
        val locationName: String,
        val temperatureC: Double,
        val condition: String,
        val advice: String?,
        val observedAtMillis: Long
    )

    private fun makeKey(lat: Double, lon: Double): String {
        // Round to 2 decimal places (~1.1 km) for regional weather cache key
        val latRound = String.format(java.util.Locale.US, "%.2f", lat)
        val lonRound = String.format(java.util.Locale.US, "%.2f", lon)
        return "weather_${latRound}_${lonRound}"
    }

    fun saveObservation(
        lat: Double,
        lon: Double,
        locationName: String,
        temperatureC: Double,
        condition: String,
        advice: String?,
        observedAtMillis: Long = System.currentTimeMillis()
    ) {
        val key = makeKey(lat, lon)
        val observation = CachedObservation(
            locationName = locationName,
            temperatureC = temperatureC,
            condition = condition,
            advice = advice,
            observedAtMillis = observedAtMillis
        )
        memoryStore[key] = observation

        prefs?.edit()?.apply {
            putString("${key}_name", locationName)
            putString("${key}_temp", temperatureC.toString())
            putString("${key}_cond", condition)
            putString("${key}_advice", advice ?: "")
            putLong("${key}_time", observedAtMillis)
            apply()
        }
    }

    fun getCachedObservation(
        lat: Double,
        lon: Double,
        nowMillis: Long = System.currentTimeMillis()
    ): WeatherState? {
        val key = makeKey(lat, lon)
        val mem = memoryStore[key]
        if (mem != null) {
            val relativeTime = formatRelativeTimeAgo(mem.observedAtMillis, nowMillis)
            return WeatherState.Cached(
                locationName = mem.locationName,
                temperatureC = mem.temperatureC,
                condition = mem.condition,
                advice = mem.advice,
                relativeTimeAgo = relativeTime
            )
        }

        val p = prefs ?: return null
        if (p.contains("${key}_time")) {
            val name = p.getString("${key}_name", "Odisha") ?: "Odisha"
            val tempStr = p.getString("${key}_temp", null) ?: return null
            val temp = tempStr.toDoubleOrNull() ?: return null
            val cond = p.getString("${key}_cond", "Clear") ?: "Clear"
            val adv = p.getString("${key}_advice", null)?.takeIf { it.isNotBlank() }
            val time = p.getLong("${key}_time", nowMillis)

            val observation = CachedObservation(name, temp, cond, adv, time)
            memoryStore[key] = observation

            val relativeTime = formatRelativeTimeAgo(time, nowMillis)
            return WeatherState.Cached(
                locationName = name,
                temperatureC = temp,
                condition = cond,
                advice = adv,
                relativeTimeAgo = relativeTime
            )
        }

        return null
    }

    fun formatRelativeTimeAgo(observedAtMillis: Long, nowMillis: Long): String {
        val diffMillis = (nowMillis - observedAtMillis).coerceAtLeast(0L)
        val diffMinutes = diffMillis / (1000 * 60)
        val diffHours = diffMinutes / 60
        val diffDays = diffHours / 24

        return when {
            diffMinutes < 1 -> "just now"
            diffMinutes < 60 -> "${diffMinutes}m"
            diffHours < 24 -> "${diffHours}h"
            else -> "${diffDays}d"
        }
    }

    companion object {
        @Volatile
        private var INSTANCE: WeatherCacheStore? = null

        fun getInstance(context: Context?): WeatherCacheStore {
            return INSTANCE ?: synchronized(this) {
                INSTANCE ?: WeatherCacheStore(context?.applicationContext ?: context).also {
                    INSTANCE = it
                }
            }
        }

        fun createMock(): WeatherCacheStore = WeatherCacheStore(null)
    }
}
